module/bdev/lvol/vbdev_lvol.c

```diff
@@ -934,43 +1050,146 @@ lvol_write(struct spdk_lvol *lvol, struct spdk_io_channel *ch, struct spdk_bdev_
 static int
 lvol_reset(struct spdk_bdev_io *bdev_io)
 {
-	spdk_bdev_io_complete(bdev_io, SPDK_BDEV_IO_STATUS_FAILED);
+	lvol_op_comp(bdev_io, -EPERM);
 
 	return 0;
 }
 
 static void
 lvol_get_buf_cb(struct spdk_io_channel *ch, struct spdk_bdev_io *bdev_io, bool success)
 {
+	struct spdk_lvol_channel *lvol_ch = (struct spdk_lvol_channel *)spdk_io_channel_get_ctx(ch);
+
 	if (!success) {
 		spdk_bdev_io_complete(bdev_io, SPDK_BDEV_IO_STATUS_FAILED);
 		return;
 	}
 
-	lvol_read(ch, bdev_io);
+	lvol_read(lvol_ch->bs_channel, bdev_io);
+}
+
+static void
+lvol_flush_unquiesce_cb(void *cb_arg, struct spdk_lvol *lvol, int lvolerrno)
+{
+	struct spdk_bdev_io *bdev_io = cb_arg;
+
+	lvol_op_comp(bdev_io, 0);
+}
+
+static void
+lvol_flush_quiesce_cb(void *cb_arg, struct spdk_lvol *lvol, int lvolerrno)
+{
+	struct spdk_bdev_io *bdev_io = cb_arg;
+	int rt;
+
+	if (lvolerrno != 0) {
+		lvol_op_comp(bdev_io, lvolerrno);
+	}
+
+	rt = _vbdev_lvol_unquiesce(lvol, bdev_io, lvol_flush_unquiesce_cb, bdev_io);
+	if (rt != 0) {
+		lvol_op_comp(bdev_io, rt);
+	}
+}
+
+static void
+lvol_flush(struct spdk_bdev_io *bdev_io)
```

Comment:
I'm curious about why the upstream doesn't provide flush function. Do you know the context?

Reply:
No, I don't know the reason why lvol didn't implement the flush command. But after having introduced the quiesce functionality into bdev layer, the lvol implementation has been easier because it basically replicate that way of doing.